### PR TITLE
[JENKINS-28180] The fluid layout has been improved to support long job names

### PR DESCRIPTION
**Pull request by recena**
https://issues.jenkins-ci.org/browse/JENKINS-28180

## Before
jenkins-28180_before

## After
jenkins-28180_before

@reviewbybees, specially @tfennelly

### CODE REVIEW

File: src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView/bpp.jelly [hunk 1]
```diff
@@ -365,19 +367,23 @@
 									<!-- Only adding the revision-cell once. -->
 									<j:if test="${x == 0}"><td class="revision-cell" rowspan="${buildGrid.rows}">
 										<table class="revision rounded build-card">
-											<tbody>
-												<tr class="header">
```
Comment (oleg-nenashev) [inline src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView/bpp.jelly]: What was the purpose of moving/removing existing classes? If somebody has a custom CSS theme, this change will break them

File: src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView/bpp.jelly [hunk 2]
```diff
@@ -17,10 +17,10 @@
 	<script src="${rootURL}/plugin/build-pipeline-plugin/js/build-pipeline.js"></script>
 	<script id="build-card-template" type="text/x-handlebars-template">
 <table class="build-card rounded {{build.status}} {{#if project.disabled}}disabled{{/if}}">
-	<tbody>
-		<tr class="header">
+	<thead>
+		<tr>
 			<td>
-				<div class="overflow-hidden">
+				<div class="header">
```
Comment (oleg-nenashev) [inline src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView/bpp.jelly]: Unrequired class renaming

**Comment by oleg-nenashev** [comment]
Is there a way to get a full name from the UI? (e.g. a popup when a mouse hovers the text)

File: src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView/bpp.jelly (hunk 2 above)
Comment (recena) [inline src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView/bpp.jelly]: @oleg-nenashev Please, review the changes. It is not only a CSS class rename. The semantic is completely different.

**Comment by oleg-nenashev** [comment]
I don't understand the purpose of the most of changes + I'm aware about regressions in users CSS styles. Probably, I just don't have enough experience in Web development => I abstain

File: src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView/bpp.jelly (hunk 1 above)
Comment (recena) [inline src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView/bpp.jelly]: @oleg-nenashev This is a horrible **bad-practice**. If we assume this, we will never modify any CSS style or HTML DOM.

Comment (oleg-nenashev) [inline src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView/bpp.jelly]: A custom CSS theme has been a common approach to Jenkins customization for years. Why do you think that it is a bad practice? What is the good practice then?

**Comment by recena** [comment]
@oleg-nenashev I only tried to implement `text-overflow` in a right way. The previous code does not work fine. In an UI implemented using HTML & CSS is very normal to have to do these kind of changes. You need to change the DOM and CSS.

**Comment by oleg-nenashev** [comment]
The new way works properly at least, so the change is reasonable in any case :)
I just don't feel convenient enough to elaborate the impact of changes.

**Comment by recena** [comment]
@oleg-nenashev We could release a `beta version` including this PR and others related to UI.

File: src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView/bpp.jelly (hunk 1 above)
Comment (recena) [inline src/main/resources/au/com/centrumsystems/hudson/plugin/buildpipeline/BuildPipelineView/bpp.jelly]: @oleg-nenashev The bad practice is not to use custom CSS overriding the default implementation but to use not semantic pseudo classes, etc...

As I said in other comments, we have to do an effort to improve this part of Jenkins project. It spend much time about binary compatibility, API compatibility, etc... but the UI is a very important part of the code.

**Comment by tfennelly** [comment]
> Is there a way to get a full name from the UI? (e.g. a popup when a mouse hovers the text)

I can see that the project name is set on the `title` attribute of the `<a>` inside the `header`, which means the full project name would be shown on mouse over.

**Comment by tfennelly** [comment]
I agree with @recena wrt how the content is currently marked up using class names like `overflow-hidden` etc. That is bad practice. Also, how the CSS was applied globally to elements having that class name ... that's not ideal either.

So, I think the changes made by @recena probably do improve things, but I also hear what @oleg-nenashev is saying wrt custom CSS themes.

On the other hand, if we are saying that UI elements should never change for fear of breaking some custom theme, then I think we are effectively saying that we can never change UI, which would be crazy imo. In the case of this PR, it's not just the class name changes that could break a custom theme; the structural changes could too.

So:

1. Is there a Jenkins policy around guaranteeing backward compatibility for custom themes? I'm not aware of one.
1. In the case of this plugin, do we know of any custom themes that people might have that can't easily be "fixed" to work against the new layout.

**Comment by recena** [comment]
@tfennelly As you know well, there are UI issues that require changes both HTML DOM & CSS. If we try to solve all issue without modifications, probably the result will be **spaghetti code** :tired_face: 

For that, we (Jenkins community) should be opened to these kind of changes.

**Comment by tfennelly** [comment]
:bee: 

I agree this may cause issues for anyone using a custom theme, but then any UI change has that associated risk.

**Comment by tfennelly** [comment]
... and I think anyone using a custom theme has already accepted that risk by using it in the first place.

**Comment by recena** [comment]
@tfennelly The themes (like the rest of the code) is alive

**Comment by tfennelly** [comment]
@recena well I think they are kinda different tbh ... to me they are like a parallel universe and if you decide to go there then you need to be OK with things not being so great when things change back on the other side.

**Comment by recena** [comment]
@tfennelly I agree but if the changes involves **removing** bad-practices, go ahead!

**Comment by tfennelly** [comment]
@recena sure ... there's no "but" there ... we are saying the same thing really.

**Comment by recena** [comment]
@tfennelly you know, my bad english...

**Comment by recena** [comment]
@reviewbybees done

**Comment by tfennelly** [comment]
@recena your English is great Manuel (infinitely better than my Spanish ;) )

**Comment by KostyaSha** [comment]
:-1: just cutting names is not a solution

**Comment by recena** [comment]
@KostyaSha Any constructive reason? Please, review the result and the implementation.

**Comment by KostyaSha** [comment]
I see in screenshot that it cuts job names. Plugin firstly created for displaying activities and what sense to use it if they are cut?

**Comment by KostyaSha** [comment]
On what jenkins versions have you verified it?

**Comment by recena** [comment]
@KostyaSha 

1. The [previous code](https://github.com/jenkinsci/build-pipeline-plugin/blob/build-pipeline-plugin-1.4.7/src/main/webapp/css/main.css#L85) *tried* to cut the title. Did not work fine.
1. The current implementation is based on a liquid layout in order to use the screen width.

If you find a bug, please, file a JIRA ticket and I'll solve it.

**Comment by KostyaSha** [comment]
> This pull request originates from a CloudBees employee.

+
> we (Jenkins community)

Nice!

**Comment by recena** [comment]
@KostyaSha Could you provide any technical argument to this PR? Since you participated on this PR I did not received any contribution.

**Comment by KostyaSha** [comment]
@recena have you tried variant where letters will be moved on the next line? This will allow showing real names. I also think that if all jobs will be places in long folders path you wouldn't understand what job is running now (don't remember exactly how plugin shows jobs in folders, but AFAIR something was fixed).

**Comment by KostyaSha** [comment]
Btw your screenshot this very well demonstrates: before - you can understand  *JOB1 -> *JOB2->  cut

**Comment by recena** [comment]
@KostyaSha The first screenshot demostrate how this bug brokes the UI layout.

**Comment by KostyaSha** [comment]
In first variant i can scroll page (even if it screwed right border) and read job names. After screenshot just cuts information.

**Comment by recena** [comment]
@KostyaSha It is clear. you prefer one option and me, other. For my side, the conversation about this PR is finished. Thanks so much.

**Comment by KostyaSha** [comment]
@recena this was implemented by authors, you come and enforced your side (the same for pom formatting).

**Comment by recena** [comment]
@KostyaSha This PR was opened 10 days. Three developers (one of them, a very active contributor in Jenkins UI) have participated. I sent emails to the company and mailing list.

> 11 PR opened and the **last release** date on Mar.

This is not positive for the Jenkins Community, especially when there is people that want to do things.

**Comment by KostyaSha** [comment]
> This is not positive for the Jenkins Community, especially when there is people that want to do things.

Please say this to Centrum System and get reply from them, then you can do anything you want.

> This is not positive for the Jenkins Community, 

@daniel-beck seems it's new CB trend using this words. Could somebody define who is the "Jenkins Community"?

> (one of them, a very active contributor in Jenkins UI)

Yes, after his core changes this plugin get regressions :)

@recena Do you have any answer for fact that now it impossible to see full job name while there is enough space on the second line?

**Comment by recena** [comment]
@KostyaSha Bye bye!

**Comment by KostyaSha** [comment]
?

**Comment by tfennelly** [comment]
@KostyaSha The full text should be visible on mouse-over.

Sure, it may be possible to show it all by wrapping the text etc, but that gets messy quickly + has obvious limits + looks crap + requires JavaScript to work properly.

**Comment by KostyaSha** [comment]
@tfennelly using mouse for showing this flow on external monitor/PC is not convenient.

**Comment by dserodio** [comment]
:-1: Truncating the job names is worse than having to scroll horizontally

**Comment by recena** [comment]
@dserodio The job title was being truncated before but the current implementation was not working fine.

**Comment by dserodio** [comment]
@recena In the "before" screenshot, I can see the whole title for JOB1 and JOB2, but not JOB3. In the "after" screenshot, I can't see the whole title for *any* of the 3 jobs. Maybe I misunderstood, but isn't it possible to see the full JOB3 title in the "before" version by scrolling to the right?

**Comment by tfennelly** [comment]
@KostyaSha fair point. And of course that would be a common use case for this. I guess that also means that scrolling is of limited use too, which means wrapping has limitations.

But this text has always been truncated according to @recena.

@dserodio @recena On the original/before image ... can we demonstrate what it was truncating? Would be worth verifying how truncation was already in operation, but not working properly.

So if we verify that the text has always been truncated and that this PR is just making that truncation work properly, then where does that leave us. We're hardly going to reject a PR because it doesn't change existing behaviour to work the way we'd have wanted it to?

**Comment by recena** [comment]
@dserodio In the previous version (before) you can see the full titles in the three box and you also can see a broken layout. The result: header and footer cutted and a horizontal scroll.

We can discuss about the UX (horizontal scrolls) but the key part here is different. With this PR we have:

1. We avoid that a plugin brokes the Jenkins layout.
1. A liquid layout for the pipeline container. This was the first intension by the authors.
1. Some fixed details.

Obviously, we can choose different ways and I want to participate in all of them, but I lose my motiviation with some comments and behavior. https://groups.google.com/d/msg/jenkinsci-dev/6Eqz7Z9QbNE/TI2AwLW5AAAJ

**Comment by dserodio** [comment]
Here are 2 examples of pipelines that look much worse with the new layout:

pipeline_view_1
pipeline_view_2

**Comment by recena** [comment]
@dserodio I'd like to help here. Could you file a JIRA ticket? It would be nice if you attach these screenshots but including the browser (full screen).

**Comment by dserodio** [comment]
Ok @recena should I file a single ticket for both of these, or two separate tickets?

**Comment by dserodio** [comment]
Filed https://issues.jenkins-ci.org/browse/JENKINS-30846

**Comment by recena** [comment]
@dserodio Perfect!

**Pull request merged**